ui/app/js/directives/permission_options_directive.js

```diff
@@ -0,0 +1,127 @@
+'        <input type="text" typeahead="action for action in actions[permission.permission] | filter:$viewValue | limitTo:16" ng-model="action"></input>'+
+'    </div>'+
+'</div>'+
+'<div class="form-group" ng-show="!!(selected_options.indexOf(\'products\')+1)">'+
+'    <label><i>products</i></label>'+
+'    <div class="row-fluid m-tag__input">'+
+'        <label ng-repeat="product in selected_products" class="btn btn-sm btn-primary">{{ product }} <span class="badge" ng-click="productRemove(product)">x</span></label>'+
+'        <input type="text" typeahead="product for product in products | filter:$viewValue | limitTo:16" ng-model="product"></input>'+
+'    </div>'+
+'</div> ',
+        controller: function($scope, Releases){
+            var allPermissions = {
+                "admin": ["products"],
+                "release": ["actions", "products"],
+                "release_locale": ["actions", "products"],
+                "release_read_only": ["actions", "products"],
+                "rule": ["actions", "products"],
+                "permission": ["actions"],
+                "required_signoff": ["products"],
+                "scheduled_change": ["actions"],
+            };
+            $scope.$watch('permission.permission',function(){
+                $scope.available_options = allPermissions[$scope.permission.permission];
+
+                // clear  all previous selections 
+                $scope.selected_options = [];
+                $scope.selected_actions = [];
+                $scope.selected_products = [];
+            });
+            $scope.selected_options = [];
+            $scope.actions = {
+                "admin": [],
+                "rule": ["create", "modify", "delete"],
+                "release": ["create", "modify", "delete"],
+                "release_read_only": ["set", "unset"],
+                "release_locale": ["modify"],
+                "required_signoff": ["create", "modify", "delete"],
+                "permission": ["create", "modify", "delete"],
+                "scheduled_change": ["enact"],
+                "":[]
+            };
+
+            $scope.selected_actions = [];
+            $scope.action = null;
+
+            $scope.products = [];
+            $scope.selected_products = [];
+            $scope.product = null;
+
+            // get products in releases
+            Releases.getProducts().success(function(response) {
+                $scope.products = response.product;
+            });
+
+            $scope.$watch('[ action, product ] | json',function(){
+                // add tag on action input
+                if ($scope.actions[$scope.permission.permission].indexOf($scope.action) !== -1 && $scope.selected_actions.indexOf($scope.action) === -1){
+                    $scope.selected_actions.push($scope.action);
+                    $scope.action = null;
```

Comment:
Have you considered using sets (https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set) for the `selected` variables? They should let you avoid doing duplication checks manually.

Reply:
Yeah, will also do that ASAP

Reply:
Tried using sets but I keep on getting `Error: Can't find variable: Set` when running tests

Reply:
I think we talked about this elsewhere...sets probably aren't available in the version of phantomjs that we use, we'll have to do without them :(

Reply:
Yeah 👍